### PR TITLE
Add docker version info in /info request

### LINKED ISSUE #14952: Remote API : add docker's version information in info request
Is it possible to add docker's version in result for /info request for remote API ?

Thanks in advance.
Regards.

### PR DESCRIPTION
**Pull request by azurezk**
fix #14952
Signed-off-by: Zhang Kun zkazure@gmail.com

### CODE REVIEW

**Comment by jfrazelle** [comment]
I dont know if this is something we need we have the /version endpoint already :/

**Comment by jfrazelle** [comment]
ping @docker/core-maintainers

File: api/client/info.go [hunk 1]
```diff
@@ -104,5 +104,16 @@ func (cli *DockerCli) CmdInfo(args ...string) error {
 		fmt.Fprintf(cli.out, "Experimental: true\n")
 	}
 
+	fmt.Fprintf(cli.out, "Docker Version Infomation: \n")
```
Comment (vieux) [inline api/client/info.go]: `Information`

Comment (jfrazelle) [inline api/client/info.go]: @vieux is this saying you want this? I personally think its unecessary

Comment (Shini31) [inline api/client/info.go]: Hi,

2 requests are necessary if you want global information on Docker.
For example in a monitoring script two requests can raise the execution time.
Moreover there are already some duplicate information as the kernel version.

Regards.

**Comment by LK4D4** [comment]
I'm okay with adding version to info
ping @docker/core-maintainers

**Comment by thaJeztah** [comment]
It would sometimes reduce the number of going back and forth to get the information when people are reporting issues here, but either way is fine to me

**Comment by jfrazelle** [comment]
ping @icecrime @tiborvass idk im kinda neutral but leaning towards why not just call the other endpoint.... idk

**Comment by azurezk** [comment]
I updated the code.  @jfrazelle  do you mean this?

**Comment by calavera** [comment]
LGTM

**Comment by duglin** [comment]
I'm -0 on this because this will leave us with two commands that share output ; one is a direct subset of the other.  Why not just look to merge them into one command and kill off the other?

**Comment by jfrazelle** [comment]
ok yeah i agree with @duglin

**Comment by abronan** [comment]
Agreed with @jfrazelle and @duglin. This will render `docker version` virtually useless. I'm ok with adding version to info but this should be a single field with the server version for example and not just adding the complete output of `version` right after `info`.

What about something like this:
```
Containers: 0
Images: 107
Server Version: 1.8.1
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.5-1-ARCH
Operating System: Arch Linux
CPUs: 8
Total Memory: 15.6 GiB
Name: archbox
ID: OUQD:OBQE:7KQO:2FQ5:YJIC:ILXS:O7SE:PWEK:3ZWT:PG7I:RUPA:2EIW
```

**Comment by Shini31** [comment]
Agreed with @abronan 
This is exactly what I asked in #14952 (may be a misunderstanding, my english sucks ...) :)

**Comment by jfrazelle** [comment]
ok then lets move ahead with that

**Comment by jfrazelle** [comment]
@azurezk can you update this PR to do that instead?

**Comment by jfrazelle** [comment]
oh jk that is how it is now, LGTM

**Comment by jfrazelle** [comment]
ping @abronan ;) its now just client side

**Comment by azurezk** [comment]
Should it be changed on daemon side? I did it before, shall I change it back?

**Comment by cpuguy83** [comment]
ping @abronan So this should be done daemon side?

**Comment by abronan** [comment]
@jfrazelle what do you mean by it's just client side now? :) 

Imho this would be better if done daemon side by adding a single `DockerVersion` field to `/info` and use it client side on `CmdInfo`.

**Comment by icecrime** [comment]
@abronan Can you please take care of that one (i.e., review if it is as you expect, or carry if you feel it need changes)? Seems this PR has been waiting long enough.

**Comment by abronan** [comment]
@icecrime sure

@azurezk Sorry for the delay on discussing the design, are you open to make the change as it was first (ie. not just appending `version` to `info` but adding a field for the version like `DockerVersion` and output it on `info`)? Thanks!

**Comment by azurezk** [comment]
yes, code is updated:) @abronan

**Comment by abronan** [comment]
Thanks @azurezk. Failures on Jenkins appears to be unrelated. Will check later though once this is updated.

LGTM

Might need some change in the documentation for the commandline output though.

File: api/client/info.go [hunk 2]
```diff
@@ -35,6 +35,7 @@ func (cli *DockerCli) CmdInfo(args ...string) error {
 
 	fmt.Fprintf(cli.out, "Containers: %d\n", info.Containers)
 	fmt.Fprintf(cli.out, "Images: %d\n", info.Images)
+	fmt.Fprintf(cli.out, "Docker Server Version: %s\n", info.ServerVersion)
```
Comment (cpuguy83) [inline api/client/info.go]: Nit, can probably get rid of "Docker" here.

Comment (cpuguy83) [inline api/client/info.go]: Could also be `Engine Version`

**Comment by cpuguy83** [comment]
Can you add the new field to https://github.com/docker/docker/blob/master/integration-cli/docker_api_info_test.go#L18

**Comment by azurezk** [comment]
doc and test changed a little  @abronan @cpuguy83

**Comment by cpuguy83** [comment]
LGTM

**Comment by thaJeztah** [comment]
Thanks @azurezk, I think there's a few more docs that need to be updated;

- The `/info` endpoint of the API will also include this information, can you add the new information to the example here: [docs/reference/api/docker_remote_api_v1.21.md#display-system-wide-information](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.21.md#display-system-wide-information)

- Can you add a note to the "changelog" for the API? You can find that here: [docs/reference/api/docker_remote_api.md#v121-api-changes](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v121-api-changes)

- The new output also needs to be added to the example output in the man-page for `docker info`, which can be found here: [man/docker-info.1.md](https://github.com/docker/docker/blob/master/man/docker-info.1.md#display-docker-system-information)

File: docs/reference/commandline/info.md [hunk 3]
```diff
@@ -21,6 +21,7 @@ For example:
     $ docker -D info
     Containers: 14
     Images: 52
+    Engine Version: 1.8.0
```
Comment (thaJeztah) [inline docs/reference/commandline/info.md]: Just a "nit": could you use `1.9.0` in the examples (given that this feature won't be available until docker 1.9.0)

**Comment by azurezk** [comment]
docs added :) @thaJeztah

**Comment by Shini31** [comment]
Thanks for your work @azurezk !

**Comment by thaJeztah** [comment]
Thanks @azurezk 

LGTM

ping @moxiegirl @SvenDowideit

**Comment by abronan** [comment]
Sorry @azurezk, It needs a rebase :disappointed:

**Comment by abronan** [comment]
kindly ping @moxiegirl @SvenDowideit :dancer: :whale:

**Comment by moxiegirl** [comment]
Thank you @azurezk.  LGTM @thaJeztah for the merge.

**Comment by abronan** [comment]
@azurezk Needs a last rebase and we are good to merge :+1:

File: docs/reference/api/docker_remote_api.md [hunk 4]
```diff
@@ -88,6 +88,10 @@ This section lists each version from latest to oldest.  Each listing includes a
 list of DNS options to be used in the container.
 * `POST /build` now optionally takes a serialized map of build-time variables.
 * `GET /events` now includes a `timenano` field, in addition to the existing `time` field.
+* `GET /info` now lists engine version information.
+
+>>>>>>> aaeb7ad... add version info in GET /info
+>>>>>>> 62a200e... add version info in GET /info
```
Comment (runcom) [inline docs/reference/api/docker_remote_api.md]: bad rebase here?

**Comment by vieux** [comment]
:-1: could we please hide those field in the CLI if they are empty ? 
Please keep in mid that the cli doesn't always talk to an engine, sometimes it's swarm :D

**Comment by tiborvass** [comment]
@vieux if it's an important fix for 1.9, please send a PR quick so we can have it for rc2.

**Comment by vieux** [comment]
It's not "important" but I will

**Comment by vieux** [comment]
thanks @tiborvass

**Pull request merged**